### ISSUE TITLE
docs: document the unsloth surface from PraisonAI #4370 (13 new config keys + 23 new quants)

### ISSUE THREAD

**Issue by MervinPraison** [MervinPraison/PraisonAIDocs#3363]
## Summary

[PraisonAI PR #4370](https://github.com/MervinPraison/PraisonAI/pull/4370) (merged 2026‑08‑25, branch `feat-unsloth-surface`) landed a large expansion of the `praisonai-train llm` config surface — every one of these is a keyword that Unsloth **already** accepts, that the trainer was **not** forwarding, and that is **not** covered anywhere in `MervinPraison/PraisonAIDocs`. Runs that set them today would previously have printed an "unknown config key" warning and been silently narrower than the user asked for.

The additions land in `src/praisonai-train/praisonai_train/train/llm/trainer.py` (+89 / −4). Nothing was deprecated — all new keys are additive and previously-documented keys still work.

We need one new feature page and edits to five existing pages.

**Non-goal for the follow-up doc agent:** do NOT rewrite `docs/concepts/` — per `AGENTS.md`, that folder is human-approved-only. All new/updated content lands under `docs/features/`, `docs/train.mdx`, and `docs/cli/train.mdx`.

---

## Ground-truth SDK reference (read BEFORE writing)

Read these paths in `MervinPraison/PraisonAI` at merge SHA `8dd0f3310ae5f5d33e3409e96319ea5a0518e523` (or `main` post-merge) as the source of truth for every parameter, type, and default. Do not guess:

- `src/praisonai-train/praisonai_train/train/llm/trainer.py`
  - `model_access_kwargs()` (top of file) — assembles `hf_token` → `token`, `trust_remote_code`, `revision`, `fast_inference`, `gpu_memory_utilization`, `max_lora_rank`.
  - `ALLOWED_QUANTS` (24 quants) and `IMATRIX_QUANTS` (11 IQ quants) — full new quant lists.
  - `TrainModel.KNOWN_KEYS` — all newly-accepted config keys added to `load_config`.
  - `TrainModel.prepare_model` — the extended PEFT passthrough loop for `finetune_last_n_layers`, `layers_to_transform`, `layers_pattern`, `target_parameters`, `init_lora_weights`.
  - `TrainModel.load_model` — `load_kwargs.update(model_access_kwargs(...))` is called for GRPO too.
  - `TrainModel.save_merged_locally()` — new: writes merged weights to `merged_save_dir` (default `save_method` = `"merged_16bit"`).
  - `TrainModel.push_model_gguf()` — now forwards `imatrix_file` when set.
  - `TrainModel.run()` — `self.save_merged_locally()` is always called (no-op when `merged_save_dir` unset).
- `src/praisonai-train/tests/unit/train/test_unsloth_surface.py` — canonical minimum-example matrix for every new key.

The mapping (config key → what Unsloth receives):

| Config key                | Unsloth kwarg                        | Notes |
|---------------------------|--------------------------------------|-------|
| `hf_token`                | `token`                              | Renamed to avoid colliding with the trainer's own `token` handling. |
| `trust_remote_code`       | `trust_remote_code`                  | Required for custom-code models. |
| `revision`                | `revision`                           | Pin a HF revision / branch / SHA for a reproducible multi-day run. |
| `fast_inference`          | `fast_inference=True` + companions   | Turns on vLLM rollouts for GRPO. |
| `gpu_memory_utilization`  | `gpu_memory_utilization`             | Only forwarded when `fast_inference` is truthy. |
| `max_lora_rank`           | `max_lora_rank`                      | Only forwarded when `fast_inference` is truthy. |
| `finetune_last_n_layers`  | `FastLanguageModel.get_peft_model`   | Cheap-run recipe — only tune the last N transformer blocks. |
| `layers_to_transform`     | `get_peft_model`                     | Explicit layer indices. |
| `layers_pattern`          | `get_peft_model`                     | Regex/pattern for layer selection. |
| `target_parameters`       | `get_peft_model`                     | Required to LoRA MoE experts. |
| `init_lora_weights`       | `get_peft_model`                     | Init scheme (e.g. `"gaussian"`, `"pissa"`, `"olora"`, etc.). |
| `merged_save_dir`         | `model.save_pretrained_merged(target, tokenizer, save_method=…)` | Offline merge; no Hub needed. |
| `imatrix_file`            | `push_to_hub_gguf(..., imatrix_file=…)` | Only forwarded when set; unsloth otherwise auto-fetches an imatrix. |

New quant lists to mirror **verbatim** from `trainer.py`:

- `ALLOWED_QUANTS` (24): `bf16`, `f16`, `f32`, `fast_quantized`, `not_quantized`, `q2_k`, `q2_k_l`, `q3_k_l`, `q3_k_m`, `q3_k_s`, `q3_k_xs`, `q4_0`, `q4_1`, `q4_k`, `q4_k_m`, `q4_k_s`, `q5_0`, `q5_1`, `q5_k`, `q5_k_m`, `q5_k_s`, `q6_k`, `q8_0`, `quantized`.
- `IMATRIX_QUANTS` (11): `iq1_m`, `iq1_s`, `iq2_m`, `iq2_s`, `iq2_xs`, `iq2_xxs`, `iq3_m`, `iq3_s`, `iq3_xxs`, `iq4_nl`, `iq4_xs`.
- `VALID_QUANTIZATION_METHODS` = `ALLOWED_QUANTS | IMATRIX_QUANTS` (35 total).

The IQ family is what lets a 30B model fit on a laptop; it was previously unreachable — call this out in the copy.

---

### 1. NEW page — `docs/features/praisonai-train-unsloth-surface.mdx`

One user-facing page that pulls together the surface changes as a discoverable landing spot. Follow `AGENTS.md` section 2 (page structure) and section 3 (mermaid colour scheme).

Frontmatter:
```yaml
---
title: "Model access, PEFT selectors & offline merge"
sidebarTitle: "Unsloth Surface"
description: "Gated models, custom-code models, MoE LoRA, offline merged export, and vLLM rollouts for GRPO"
icon: "sliders"
---
```

Sections (each opens with an Agent-centric single-line example, then a mermaid diagram, then a `<Steps>` block — mirror the style of `docs/features/train-reward-functions.mdx`):

1. **Model access** — `hf_token`, `trust_remote_code`, `revision`. Show a gated-model example (Llama or Gemma) using `hf_token`, and one pinning `revision` for reproducibility. Note: `hf_token` in the config is the **config-level** equivalent of the `HF_TOKEN` env var already documented in `docs/features/praisonai-train-hub-privacy.mdx` — link both.
2. **PEFT selectors** — `finetune_last_n_layers` (add a "cheap run" mermaid decision diagram: full LoRA vs last-N vs MoE experts), `target_parameters` (MoE experts — a Mixtral-shaped example), `layers_to_transform`, `layers_pattern`, `init_lora_weights`.
3. **Offline merged export** — `merged_save_dir` + `save_method`. Show why it matters: "no Hub account, no token, no network — merge into a folder on this disk." Contrast with the existing Hub-push flow in `praisonai-train-export.mdx`.
4. **vLLM fast rollouts (GRPO)** — `fast_inference`, `gpu_memory_utilization`, `max_lora_rank`. Include a note that GRPO previously always ran the slow `HF-generate` path; a one-liner turns on the fast rollout.
5. **Expanded quant list** — the 35-quant table split into standard (24) and imatrix (11), plus `imatrix_file`. Include a mermaid decision diagram "which quant should I pick" (rough tiers: size vs quality vs laptop-fit).

Include a `Related` `CardGroup` linking `train`, `cli/train`, `features/praisonai-train-export`, `features/praisonai-train-hub-privacy`, `features/train-reward-functions`.

Add to `docs/docs.json` under the **Features** group — never under Concepts (per `AGENTS.md` §1.8).

### 2. UPDATE — `docs/features/praisonai-train-export.mdx`

The `## Valid --quant values` section currently lists only the old 12 quants:

```
q2_k  q3_k_m  q4_0  q4_1  q4_k_m  q5_0  q5_1  q5_k_m  q6_k  q8_0  f16  bf16
```

Replace with the full 35 (24 standard + 11 imatrix). Split visually into two subsections ("Standard" and "IMatrix (large-model / laptop-fit)"). Also add `imatrix_file` to the "Hub push options" table with type `str`, default `—`, description "Path to a precomputed importance matrix; forwarded to `push_to_hub_gguf`. Only sent when set — Unsloth auto-fetches one from `unsloth/<base>-GGUF` otherwise." Cross-link to the new unsloth-surface page.

### 3. UPDATE — `docs/features/praisonai-train-hub-privacy.mdx`

Add `hf_token` (config key) alongside the existing `HF_TOKEN` env-var docs. State the precedence: env var wins, then cached `huggingface-cli login`, then `hf_token` from config. Add a `Note` that config-level `hf_token` is preferred inside CI configs where you don't want to leak the token into environment inspection.

### 4. UPDATE — `docs/train.mdx` (largest edit)

Five separate touches are required — this page is currently 1725 lines and the changes anchor at known lines.

**4a. Replace every stale 12-quant list** (5 occurrences). The line "The 12 valid `quantization_method` values: `q4_k_m`, `q5_k_m`, `q8_0`, `q4_0`, `q4_1`, `q5_0`, `q5_1`, `q3_k_m`, `q6_k`, `f16`, `bf16`, `q2_k`." appears at approximately lines **233, 235, 1321, 1378, 1506**. Every one must be replaced with the new 35-quant surface (24 Standard + 11 IMatrix). Prefer moving the canonical list into ONE table and having the other four sites link to it, rather than duplicating a 35-item list five times.

**4b. Gated base-model access.** The existing `### Hugging Face token` section (~line 501) frames `HF_TOKEN` purely as a **publishing** credential ("Push a fine-tuned model to Hugging Face and Ollama"). This is now wrong for gated bases — `HF_TOKEN` / config `hf_token` is *also* needed to *download* Llama/Gemma weights. Rewrite this section to cover both roles, and add a two-line example that trains a gated Llama base:

```yaml
model_name: "meta-llama/Meta-Llama-3.1-8B-Instruct"
hf_token: "hf_xxx"   # or export HF_TOKEN — same token, either channel
```

**4c. New "Model access & reproducibility" subsection** near the model-loading content, adding `trust_remote_code` and `revision` (neither currently appears anywhere in the file — grep-verified 0 hits).

**4d. Extend the existing PEFT/LoRA content.** The `#3274` config-keys table at ~line 1066–1073 documents `lora_r`, `lora_alpha`, `lora_target_modules`, `lora_dropout`, `lora_bias`, `use_rslora`; the Advanced LoRA subsection at ~line 1151 has `modules_to_save`, `rank_pattern`, `alpha_pattern`, `use_dora`. Append `finetune_last_n_layers`, `layers_to_transform`, `layers_pattern`, `target_parameters`, `init_lora_weights` to the Advanced LoRA subsection with one row each.

**4e. New "Save merged locally" subsection** referring to `merged_save_dir` (the file has zero `save_pretrained` hits today — the local-directory merge flow is entirely absent).

**4f. New "Fast rollouts for GRPO" subsection** referring to `fast_inference`, `gpu_memory_utilization`, `max_lora_rank`. Also update the preference-tuning list (currently `sft`, `dpo`, `orpo`, `kto` only at ~line 1097) to add `grpo` — GRPO is not mentioned anywhere in `train.mdx` today.

Each new subsection is a two-to-four-liner + link to the new unsloth-surface page. Do **not** duplicate the deep dive.

### 5. UPDATE — `docs/cli/train.mdx`

Add each new config key to whichever config-key table this page carries (there is a KNOWN_KEYS-shaped table). Sort/group by area: model access, PEFT, quant, vLLM. If the page carries a "config.yaml full example" block, extend the commented example to include (at minimum) `hf_token`, `revision`, `merged_save_dir`.

### 6. UPDATE — `docs/features/train-reward-functions.mdx`

Add a **Fast rollouts (`fast_inference`)** block to the GRPO section, showing that flipping `fast_inference: true` swaps HF-generate for vLLM. Include the two companion knobs `gpu_memory_utilization` and `max_lora_rank`, and a sentence linking to the new unsloth-surface page.

---

## Code samples the follow-up agent MUST include

All samples user-first (per `AGENTS.md` §6): Agent-centric first, then config-level, then Python-API level. Import friendly: `from praisonaiagents import Agent`.

### Gated model + revision pin (agent-facing snippet)

```yaml
# config.yaml
model_name: "meta-llama/Meta-Llama-3.1-8B-Instruct"
hf_token: "hf_xxx"          # or leave unset and use HF_TOKEN
revision: "main"             # pin so a multi-day run is reproducible
trust_remote_code: false
```

### Cheap LoRA — last N layers

```yaml
model_name: "unsloth/Qwen2.5-32B-Instruct-bnb-4bit"
finetune_last_n_layers: 8
```

### MoE expert LoRA

```yaml
model_name: "unsloth/mistral-8x7b-instruct-v0.3-bnb-4bit"
target_parameters:
  - "mlp.experts"          # exact selector belongs in copy — read Unsloth's docs
```

### Offline merged export (no Hub)

```yaml
model_name: "unsloth/gemma-2-2b-it-bnb-4bit"
merged_save_dir: "./merged-model"
save_method: "merged_16bit"   # default; also "merged_4bit_forced", "lora"
```

### Fast rollouts for GRPO

```yaml
method: grpo
model_name: "unsloth/gemma-2-2b-it-bnb-4bit"
fast_inference: true
gpu_memory_utilization: 0.6
max_lora_rank: 32
num_generations: 8
reward_funcs:
  - myproject.rewards:length_penalty
```

### Imatrix quant (30B on a laptop)

```yaml
quantization_method: iq4_xs      # imatrix quant — auto-fetches matrix
# imatrix_file: "./my.imatrix"   # optional; supply your own
```

---

## Mermaid diagrams to author

Use the standard palette from `AGENTS.md` §3.1 (Dark Red `#8B0000` inputs, Teal `#189AB4` processes, Green `#10B981` outputs, Amber `#F59E0B` decisions, Indigo `#6366F1` config).

1. **Hero — Unsloth Surface**: `graph LR` — Config → (Model access | PEFT selectors | Offline merge | vLLM rollouts | Quant) → Trained model.
2. **Which PEFT recipe?**: `graph TB` — "How much do you have to spend?" branching to full-LoRA, last-N, MoE-experts.
3. **Which quant should I pick?**: `graph TB` — quality vs size vs fit-on-laptop → standard (q4_k_m/q8_0) vs imatrix (iq4_xs/iq3_m) branch.
4. **Fast rollout flow (GRPO)**: `sequenceDiagram` — TRL → Unsloth → (HF generate | vLLM) → scored completions → gradient.
5. **Offline vs Hub merge**: `graph LR` — Trained adapter → { `save_pretrained_merged` → local dir | `push_to_hub_merged` → Hub repo }.

---

## Docs.json entry

Add the new page under the **Features** group in `docs/docs.json` (keep alphabetical / grouped-with-siblings — cluster near the existing `praisonai-train-*` entries). Never touch the **Concepts** group.

Also verify `docs/docs.json` stays valid JSON after edits.

---

## Acceptance criteria for the follow-up doc agent

- [ ] `docs/features/praisonai-train-unsloth-surface.mdx` exists, follows the `AGENTS.md` page template, uses the standard mermaid palette.
- [ ] Every one of `hf_token`, `trust_remote_code`, `revision`, `fast_inference`, `gpu_memory_utilization`, `max_lora_rank`, `finetune_last_n_layers`, `layers_to_transform`, `layers_pattern`, `target_parameters`, `init_lora_weights`, `merged_save_dir`, `imatrix_file` appears in at least one config table.
- [ ] The `Valid --quant values` block in `praisonai-train-export.mdx` lists all 35 quants, split into Standard (24) and IMatrix (11).
- [ ] All five stale 12-quant list occurrences in `docs/train.mdx` (~lines 233, 235, 1321, 1378, 1506) are removed or replaced — grep for the string `"12 valid"` should return zero hits after the edit.
- [ ] `docs/train.mdx` no longer frames `HF_TOKEN` as publish-only — the gated-base-model download case is covered.
- [ ] `docs/train.mdx` mentions `grpo` alongside `sft`, `dpo`, `orpo`, `kto` in its methods list.
- [ ] `docs/train.mdx` gains the four new subsections; each links to the new unsloth-surface page rather than duplicating.
- [ ] `docs/features/train-reward-functions.mdx` gains a `fast_inference` block.
- [ ] `docs/features/praisonai-train-hub-privacy.mdx` mentions config-level `hf_token` alongside `HF_TOKEN` env var, with precedence noted.
- [ ] `docs.json` is valid JSON and the new page is in the Features group.
- [ ] Every example is copy-paste runnable — imports included, no `your-token-here` placeholders (use realistic `hf_xxx` / `me/my-model`).
- [ ] Nothing new lands in `docs/concepts/`.

---

_Auto-filed from a webhook watching `MervinPraison/PraisonAI` PR closures; the closed PR is [#4370](https://github.com/MervinPraison/PraisonAI/pull/4370)._
